rlclientlib/extensions/onnx/src/tensor_notation.cc

```diff
@@ -0,0 +1,312 @@
+//#pragma once
+
+#include "tensor_notation.h"
+#include "err_constants.h"
+#include "api_status.h"
+
+#include <memory>
+#include <assert.h>
+
+// for base64 decoding
+#include <cpprest/http_client.h>
+
+namespace reinforcement_learning { namespace onnx {
+
+// This implements a very restricted parser/serializer for a JSON-like dialect to
+// describe input to the ONNX Runtime
+//
+// The format is:
+// <TENSORS> := "{" <TENSOR-LIST> "}"
+// <TENSOR-LIST> := <TENSOR> ["," [<TENSOR-LIST>]]
+// <TENSOR> := "'" <INPUT-NAME> "':'" <TENSOR-DATA> "'"
+// <TENSOR-DATA> := <DIMS-BASE64> ";" <VALUES-BASE64>
+// <DIMS-BASE64> := { base64 encoding of int64[] representing dimensions of the tensor }
+// <VALUES-BASE64> := { base64 encoding of float[] representing values of the tensor }
+//
+// Any other JSON concepts are not allowed. The reason for any relation to JSON is
+// the current setup for RLClientLib to log the context as JSON.
+// Ideally, we would use protobuf definitions from ONNX to represent the IOContext
+
+namespace tokens {
+  const char NULL_TERMINATOR = '\0';
+  const char ESCAPE = '\\';
+  const char SQUOTE = '\'';
+  const char SEMICOLON = ';';
+  const char COLON = ':';
+  const char COMMA = ',';
+  const char OPEN_CBRACKET = '{';
+  const char CLOSE_CBRACKET = '}';
+}
+
+using bytes = std::vector<unsigned char>;
+
+#define NO_ERROR_OR_RETURN(val) \
+  if (_has_error) { \
+    return val; \
+  }
+//!define# PARSER_HEADER
+
+#define NO_ERROR_OR_RETURN_DEFAULT(val) \
+  if (_has_error) { \
+    return val; \
+  }
+//!define# PARSER_HEADER
+
+class TensorParser
+{
+public:
+  // warnings/errors?
+  TensorParser(const char* tensor_notation, OnnxRtInputContext& input_context);
+
+  inline bool succeeded()
+  {
+    return !_has_error;
+  }
+
+  inline std::string error_message()
+  {
+    return _error_message;
+  }
+
+private:
+  inline void skip_whitespace()
+  {
+    NO_ERROR_OR_RETURN();
```

Comment:
Not sure how I feel about this error handling technique. Given that many portions of rlclientlib already throw, should this module just throw and ensure that it is caught and converted to error codes?